File: src/python/shared/operational_window.py
import yaml
import os
from datetime import datetime
import pytz

def is_operational_window_active() -> bool:
    ist = pytz.timezone("Asia/Kolkata")
    now_ist = datetime.now(ist)
    current_hour = now_ist.hour

    # Get project root
    project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))
    config_path = os.path.join(project_root, "config", "config.yaml")

    try:
        with open(config_path, "r") as f:
            config = yaml.safe_load(f)
            window = config.get("operational_window", {})
            start = window.get("start_hour_ist", 0)
            end = window.get("end_hour_ist", 24)
    except Exception:
        start = 0
        end = 24

    if start == end:
        return True # 24/7

    if start > end:
        return current_hour >= start or current_hour < end
    else:
        return current_hour >= start and current_hour < end

File: src/typescript/shared/operational-window.ts
import { DateTime } from 'luxon';
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import * as path from 'path';

export const isOperationalWindowActive = (): boolean => {
  let start = 0;
  let end = 24;

  try {
    const configPath = path.join(process.cwd(), 'config', 'config.yaml');
    const fileContents = fs.readFileSync(configPath, 'utf8');
    const config = yaml.load(fileContents) as any;
    
    if (config?.system?.operational_window) {
      start = config.system.operational_window.start_hour_ist ?? 0;
      end = config.system.operational_window.end_hour_ist ?? 24;
    }
  } catch (e) {
    console.warn('[OperationalWindow] Failed to load config, defaulting to 24/7');
  }

  const now = DateTime.utc();
  const istTime = now.setZone('Asia/Kolkata');
  const currentHour = istTime.hour;

  if (start === end) {
    return true; // 24/7
  }

  if (start === 0 && end === 24) {
    return true;
  }

  if (start > end) {
    // Over-midnight window (e.g. 21:00 to 06:00)
    return currentHour >= start || currentHour < end;
  } else {
    // Standard window (e.g. 09:00 to 17:00)
    return currentHour >= start && currentHour < end;
  }
};
